### PR TITLE
Add support for SK6812 RGBW

### PR DESCRIPTION
**Pull request by dennispg**
Adds support for SK6812 RGBW strips. #45

### CODE REVIEW

File: PixelDriver.h [hunk 1]
```diff
@@ -106,6 +106,10 @@ class PixelDriver {
     void ICACHE_RAM_ATTR show();
     uint8_t* getData();
 
+    inline bool hasWhite() {
+        return this->color_count == 4;
```
Comment (henrygab) [inline PixelDriver.h:110]: As a general engineering rule, I recommend avoiding the overloading of meaning of a variable.
Instead, I would recommend storing this value separately, perhaps initializing the variable based on the configuration's attached LED type?

File: EffectEngine.cpp [hunk 2]
```diff
@@ -147,9 +156,22 @@ bool EffectEngine::isValidEffect(const String effectName) {
...
-    _ledDriver->setRValue(idx, (uint8_t)(color.r * _effectBrightness) );
-    _ledDriver->setGValue(idx, (uint8_t)(color.g * _effectBrightness) );
-    _ledDriver->setBValue(idx, (uint8_t)(color.b * _effectBrightness) );
+    if (_ledDriver->hasWhite()) {
+        _ledDriver->setRValue(idx, (uint8_t)(color.r * _effectBrightness) );
+        _ledDriver->setGValue(idx, (uint8_t)(color.g * _effectBrightness) );
+        _ledDriver->setBValue(idx, (uint8_t)(color.b * _effectBrightness) );
+    }
+    else {
+        _ledDriver->setValue(3 * idx + 0, (uint8_t)(color.r * _effectBrightness) );
```
Comment (henrygab) [inline EffectEngine.cpp:165]: Can you help me understand why you are not always using `->set[RGB]]Value` function here?  It looks as though you've hard-coded the R/G/B ordering for non-white chipsets?

File: PixelDriver.cpp [hunk 3]
```diff
@@ -219,39 +265,46 @@ const uint8_t* ICACHE_RAM_ATTR PixelDriver::fillWS2811(const uint8_t *buff,
...
+            for (size_t led = 0; led < szBuffer / this->color_count; led++) {
                 uint16 modifier = led / cntGroup;
-                asyncdata[3 * led + 0] = pixdata[3 * modifier + 0];
-                asyncdata[3 * led + 1] = pixdata[3 * modifier + 1];
-                asyncdata[3 * led + 2] = pixdata[3 * modifier + 2];
+                for (uint8_t c = 0; c < this->color_count; c++) {
+                    asyncdata[this->color_count * led + c] = pixdata[this->color_count * modifier + c];
+                }
             }
         } else {  // Zigzag copy
```
Comment (henrygab) [inline PixelDriver.cpp:292]: At this point, it may be beneficial to (at least via comment) indicate the else case is defined as `== PixelType::GECE`.

File: PixelDriver.h (hunk 1 above)
Comment (henrygab) [reply PixelDriver.h:110]: Also, it's not clear if there's any adjustment available in this for the different white color temperatures.
What color temperature did you adjust the LED effects to support?
Generally, I would (at a minimum) recommend tracking the white temperature of each white pixel.

Also, expose this via the UI ... so it's clear that the effects were designed based on a given color 
temperature.  (e.g., 2700K?  5000K? ... it really affects the result...)

**Review by henrygab — COMMENTED** [review]
_NOTE: I am not in any way an official maintainer of this depot._

First, although I've not tested it, it seems to be well written code.

White has some quirks, especially if you want to enable strips that use different color temperature for the white LED portions.

Therefore, I did have a few questions that popped out while reviewing.  If you would help me understand the answers, that'd be appreciated!

I also had a few recommendations.  One major recommendation relates to storing the white's color temperature ... which can dramatically affect the calculations that would be used in the effects.

Final questions not related to specific lines:
If attached a white-enabled LED strip, can the use of the white portion be disabled by the user, so it acts as an RGB-only strip (at least for the built-in effects)?  If no, I would recommend this, in case the color temperature the user has does not match the one the effect was designed to use.

Again, overall solid code.  Nice.

**Comment by forkineye** [comment]
The reason I'm holding off merging this is that all the active development is taking place in the Unify branch which is in testing now.  Once Unify rolls out as 4.0, we plan on adding rgbw support.

File: PixelDriver.h (hunk 1 above)
Comment (dennispg) [reply PixelDriver.h:110]: So full disclosure, I am totally an amateur with this stuff. I have an RGBW strip that I wanted to drive with this code and so I figured out how to make it work and thought I would contribute it back to whomever it might benefit.

I also tried to avoid changing too much about how it currently works. `color_count` was already there, so I used it. Curious how you mean that it is overloading it's meaning though? The value is initialized in PixelDriver.cpp, line 95 based on the configuration.

I'll be honest I didn't spend too much time with the effects, except to modify them to set there is W pixel present.

File: EffectEngine.cpp (hunk 2 above)
Comment (dennispg) [reply EffectEngine.cpp:165]: so I think the ordering is translated in a later stage, so RGB must be hardcoded here..

I don't have a good answer for why I didnt just always use `->set[RGB]Value` except that I was trying to only add functionality for white without potentially adversely affecting existing functionality. Doing it the way I did here avoided the new dependency I added here on `->color_count` for non-white.

Looking at it closely there, it seems like there would be no harm to change the function to just be this:
```
        _ledDriver->setRValue(idx, (uint8_t)(color.r * _effectBrightness) );
        _ledDriver->setGValue(idx, (uint8_t)(color.g * _effectBrightness) );
        _ledDriver->setBValue(idx, (uint8_t)(color.b * _effectBrightness) );
```

File: PixelDriver.h (hunk 1 above)
Comment (dennispg) [reply PixelDriver.h:110]: Also, isnt the color temperature fixed dependending on the specific LED strip? Like I don't think you can programmatically change their color. From what I understand there are RGBCW, RGBWW and RGBNW variants for cold white, warm white and natural white, but they are all driven by SK6812 the same.

Comment (henrygab) [reply PixelDriver.h:110]: Hi Dennis,

<details><summary>Overloading of meaning</summary><P/>
This is not an industry-standard term, just prose.  Here are two articles that should help explain:
* [Separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns)
* [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single_responsibility_principle)

Thus, a single variable should correspond to a single concept.  The `color_count` indicates how many distinct LEDs per pixel need data.  This does not indicate a color ordering, nor does it indicate if any of the LEDs are white (nor the color temperature of white).

Separating it, there would be a method to uniquely identify, for each color, what color the LED in that pixel is.
</details>

<details><summary>Color Temperature Fixed?</summary><P/>

Yes, the color temperature for a given white LED is fixed at manufacture.
Yes, it's typically the same for all white LEDs in common pixel strips.
Yes, the protocol for (using your terms) RGBcW, RGBwW, RGBnW is identical, as it the driver chip (SK6812) ... it requires user input to know which color temperature, just like it needs user input to know the RGB color order.

</details>

In short, you should give yourself more credit!  You do appear to understand things quite well.  :)

File: EffectEngine.cpp (hunk 2 above)
Comment (henrygab) [reply EffectEngine.cpp:165]: Yes, having these this would be good, as it would encapsulate the offset calculation for each color into its own function.  Having a single set of functions that abstract both color order and `color_count` is a great example of how to apply the single responsibility principle, and thus will help avoid bugs later.  Nice thinking!

**Comment by forkineye** [comment]
Support is now there in the Unify branch, but it I haven't personally tested it yet as I don't have SK6812 lights yet.  Closing this and will follow up in #45
